design/design.md

```diff
@@ -0,0 +1,414 @@
+more efficient due to no lottery and less complexity, and in the decentralized
+setting, an unweighted version could be more efficient due to not needing to
+sample the binomial distribution.
+
+We should be able to reuse most of the logic between different schemes anyway,
+so the overhead shouldn't be large.
+
+### Choice of hash function
+The "random" functions $H_0$, $H_1$, $H_2$ and others in the ALBA constructions
+need to be implemented using some concrete function, either a cryptographic
+hash function modeled as a random oracle or a
+[pseudorandom function](https://en.wikipedia.org/wiki/Pseudorandom_function_family).
+There are numerous options available for one or both models:
+SHA, BLAKE, AES, ChaCha, etc.
+
+In our implementation, we will let users provide their own function.
+Hash functions evolve, old ones get broken (SHA-1), and so our ALBA
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
+Moreover, we should *require* users to provide their own hash function.
+One reason is that when ALBA is used alongside other sub-protocols, that also
+need hashing, in a larger protocol, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+Additionally, different sub-protocols should not use different hash function
+constructions (e.g., SHA and BLAKE) since there might be hidden correlations
+between them and that would be equivalent to not using domain separation.
+
+Thus, the user must use a particular hash function $H$ of their choice to
+instantiate several domain separated hash functions $H_i(x) = H(i, x)$ that
+would be provided to the sub-protocols (one of them being ALBA).
+To our ALBA implementation, this domain separation would be transparent.
+
+We will still provide an example implementing a hash function with the right
+interface for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+### Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+### Parameter calculation
```

Comment:
@djetchev  and I agreed that versioning would be suffficient for evolving the protocol so I would remove the last paragarph about backward compatibility.

As for recommendation about the derivation and hardcore, this is out of the scope and not necessarily good. We could also recommend to cache the parameters (if we use alba many times with changing parameters) for instance.

Reply:
Deleted this section.